### PR TITLE
db: heuristic to prevent tiny flushed files due to Lbase

### PR DESCRIPTION
**Pull request by sumeerbhola**
A huge Lbase, due to a compaction backlog, can currently
cause a memtable that is wide in the key space to be
flushed as many tiny files. This heuristic is an unrefined
way of preventing such a situation, but at the cost of
wide L0 files.

### CODE REVIEW

**Review by itsbilal — COMMENTED** [review]
I mostly like this approach. But I'll let others chime in on this as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1005#-:-MMWyX_3APqltR35_lTc:bu6qd66)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @itsbilal, @jbowens, @petermattis, and @sumeerbhola)

___
*[compaction.go, line 721 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1005#-MMWx6ayBq1S06BgOfz4:-MMWx6ayBq1S06BgOfz5:bbqiec0) ([raw file](https://github.com/cockroachdb/pebble/blob/114c2b9ca9f524514781f88574f14aa7cf011e03/compaction.go#L721)):*
> ```Go
> 		approxNumFilesBasedOnTargetSize :=
> 			int(math.Ceil(approxOutputBytes / float64(c.maxOutputFileSize)))
> 		acceptableFileCount := float64(4 * approxNumFilesBasedOnTargetSize)
> ```

I like the 4x adjustment to counter the pessimism. Another thing that we could consider: setting `maxOutputFileSize` to `min(flushingBytes / 2, originalMaxOutputFileSize)`. Basically guarantees that you see at least *some* flush splits no matter what, and might avoid some extreme case of extra-wide L0 files.

**Review by petermattis — COMMENTED** [review]
The approach :lgtm:. I think it is worthwhile to extract that computation into a helper routine that can be tested to ensure it is working as expected.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1005#-:-MMY2AFwCYMCz69DHDHq:b-2v10iq)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @jbowens, @petermattis, and @sumeerbhola)

___
*[compaction.go, line 717 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1005#-MMY2Lh76iWb7b5-WhNi:-MMY2Lh76iWb7b5-WhNj:b-c1124i) ([raw file](https://github.com/cockroachdb/pebble/blob/114c2b9ca9f524514781f88574f14aa7cf011e03/compaction.go#L717)):*
> ```Go
> 		// distribution of the flush. The 4x multiplier mentioned earlier is
> 		// a way to try to compensate for this pessimism.
> 		const approxCompressionRatio = 0.2
> ```

The compression ratio is highly dependent on the data. Not worth doing in this PR, but we could store in `FileMetadata` the uncompressed size next to the file size to give a good estimate of the compression ratio.

**Review by sumeerbhola — COMMENTED** [review]
TFTRs!

I've added a test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1005#-:-MMcRhcQ79DL-jpuDOJS:bfc8etp)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[compaction.go, line 717 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1005#-MMY2Lh76iWb7b5-WhNi:-MMcQmib39tw_ryQpn5c:by6nttm) ([raw file](https://github.com/cockroachdb/pebble/blob/114c2b9ca9f524514781f88574f14aa7cf011e03/compaction.go#L717)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The compression ratio is highly dependent on the data. Not worth doing in this PR, but we could store in `FileMetadata` the uncompressed size next to the file size to give a good estimate of the compression ratio.
</blockquote></details>

Good point. These are flushes so there isn't existing FileMetadata, but the other data in the key space may be representative. I've added a TODO.
___
*[compaction.go, line 721 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1005#-MMWx6ayBq1S06BgOfz4:-MMcGaRC2FDujO_fsrxm:b-kvpotj) ([raw file](https://github.com/cockroachdb/pebble/blob/114c2b9ca9f524514781f88574f14aa7cf011e03/compaction.go#L721)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I like the 4x adjustment to counter the pessimism. Another thing that we could consider: setting `maxOutputFileSize` to `min(flushingBytes / 2, originalMaxOutputFileSize)`. Basically guarantees that you see at least *some* flush splits no matter what, and might avoid some extreme case of extra-wide L0 files.
</blockquote></details>

The heuristic here is trying to reduce the effect of the grandparent based splits. If we set `c.maxOutputFileSize = min(approxCompressionRatio * flushingBytes/2, c.maxOutputFileSize)`, that doesn't seem related to the grandparents, and will just ensure that there are always at least 2 sstables written. I must be missing something.

**Review by petermattis — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1005)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @itsbilal, @jbowens, and @sumeerbhola)

___
*[compaction_test.go, line 2343 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1005#-MMfOXyX25c_oAHfjKkF:-MMfOXyX25c_oAHfjKkG:b-mo7kbk) ([raw file](https://github.com/cockroachdb/pebble/blob/d41ff60aa0b2c047225a82b4e8575671f46dd289/compaction_test.go#L2343)):*
> ```Go
> 	}
> 	for i, tc := range testCases {
> 		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
> ```

The `fmt.Sprintf("%d", i)` isn't necessary. If you pass in `""`, `t.Run` will automatically assign a serially numbered name for the test.

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1005)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[compaction_test.go, line 2343 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1005#-MMfOXyX25c_oAHfjKkF:-MMkcbXN0N_InhfnyFPh:bb6un1j) ([raw file](https://github.com/cockroachdb/pebble/blob/d41ff60aa0b2c047225a82b4e8575671f46dd289/compaction_test.go#L2343)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `fmt.Sprintf("%d", i)` isn't necessary. If you pass in `""`, `t.Run` will automatically assign a serially numbered name for the test.
</blockquote></details>

Done

**Review by itsbilal — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1005#-:-MN--zZT7KT0WmOOomp6:bnfp4nl)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @jbowens, @petermattis, and @sumeerbhola)

___
*[compaction.go, line 721 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1005#-MMWx6ayBq1S06BgOfz4:-MN--JMK37HxoKZK8M4F:bkzb8f9) ([raw file](https://github.com/cockroachdb/pebble/blob/114c2b9ca9f524514781f88574f14aa7cf011e03/compaction.go#L721)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The heuristic here is trying to reduce the effect of the grandparent based splits. If we set `c.maxOutputFileSize = min(approxCompressionRatio * flushingBytes/2, c.maxOutputFileSize)`, that doesn't seem related to the grandparents, and will just ensure that there are always at least 2 sstables written. I must be missing something.
</blockquote></details>

Makes sense. I was just suggesting that we could use another file-splitting knob (file sizes) to counter the worst-case effect of this heuristic, where worst = wide files. The particular suggestion I made would effectively put a minimum of one split (or two sstables) produced in each flush adjusted by this heuristic. Maybe that's unnecessary and I'm just overly concerned, so feel free to dismiss this suggestion.

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1005)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[compaction.go, line 721 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1005#-MMWx6ayBq1S06BgOfz4:-MN-0cZu4hSiXqSJN7gK:bsywe2i) ([raw file](https://github.com/cockroachdb/pebble/blob/114c2b9ca9f524514781f88574f14aa7cf011e03/compaction.go#L721)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Makes sense. I was just suggesting that we could use another file-splitting knob (file sizes) to counter the worst-case effect of this heuristic, where worst = wide files. The particular suggestion I made would effectively put a minimum of one split (or two sstables) produced in each flush adjusted by this heuristic. Maybe that's unnecessary and I'm just overly concerned, so feel free to dismiss this suggestion.
</blockquote></details>

Ack. I prefer not to force >= 2 sstables for each flush. We're not changing `maxOutputFileSize`, so output file size based splitting will work as usual.

**Comment by shermanCRL** [comment]
Any thoughts about backports here?

**Comment by sumeerbhola** [comment]
> Any thoughts about backports here?

waiting for approval on https://github.com/cockroachdb/pebble/pull/1139, for 20.2

**Pull request merged**
